cpp/ql/src/experimental/Security/CWE/CWE-754/ImproperCheckReturnValueScanf.ql

```diff
@@ -0,0 +1,108 @@
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-754
+ *       external/cwe/cwe-908
+ */
+
+import cpp
+import semmle.code.cpp.commons.Exclusions
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Returns the starting position of the argument being filled. */
+int posArgumentInFunctionCall(FunctionCall fc) {
+  (
+    (
+      fc.getTarget().hasGlobalOrStdName("scanf") or
+      fc.getTarget().hasGlobalOrStdName("scanf_s")
+    ) and
+    result = 1
+    or
+    (
+      fc.getTarget().hasGlobalOrStdName("fscanf") or
+      fc.getTarget().hasGlobalOrStdName("sscanf") or
+      fc.getTarget().hasGlobalOrStdName("fscanf_s") or
+      fc.getTarget().hasGlobalOrStdName("sscanf_s")
+    ) and
+    result = 2
+  )
+}
+
+from FunctionCall fc, int n
+where
+  n = posArgumentInFunctionCall(fc) and
+  // Function return value is not evaluated.
+  fc instanceof ExprInVoidContext and
+  not isFromMacroDefinition(fc) and
+  exists(Expr e0, int i |
+    i in [n .. fc.getNumberOfArguments() - 1] and
+    // Fillable argument was not initialized.
+    (
+      fc.getArgument(i).(VariableAccess).getTarget() instanceof LocalScopeVariable or
+      fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() instanceof
+        LocalScopeVariable
+    ) and
+    (
+      not fc.getArgument(i).(VariableAccess).getTarget().hasInitializer() and
+      not fc.getArgument(i)
+          .(AddressOfExpr)
+          .getOperand()
+          .(VariableAccess)
+          .getTarget()
+          .hasInitializer()
+    ) and
+    (
+      not fc.getArgument(i).(VariableAccess).getTarget().getAnAssignment().getASuccessor+() = fc and
+      not fc.getArgument(i)
+          .(AddressOfExpr)
+          .getOperand()
+          .(VariableAccess)
+          .getTarget()
+          .getAnAssignment()
+          .getASuccessor+() = fc
+    ) and
+    // After the call, the completed arguments are assigned or returned as the result of the operation of the upper function.
+    fc.getASuccessor+() = e0 and
+    (
+      (
+        e0.(Assignment).getRValue().(VariableAccess).getTarget() =
+          fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() or
+        e0.(Assignment).getRValue().(VariableAccess).getTarget() =
+          fc.getArgument(i).(VariableAccess).getTarget()
+      )
+      or
+      e0.getEnclosingStmt() instanceof ReturnStmt and
+      (
+        e0.(VariableAccess).getTarget() =
+          fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() or
+        e0.(VariableAccess).getTarget() = fc.getArgument(i).(VariableAccess).getTarget()
+      )
+      or
+      not exists(Expr e1 |
+        fc.getASuccessor+() = e1 and
+        (
+          e1.(VariableAccess).getTarget() =
+            fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() or
+          e1.(VariableAccess).getTarget() = fc.getArgument(i).(VariableAccess).getTarget()
+        )
+      )
```

Comment:
you are right this part is artificially set separately.
the criterion for the lack of evaluation of the use of the result in the comparison operation will reduce the number of detections.

this criterion is not an error condition, rather, on the contrary, if you did not initialized a variable, did not check the correctness of the call, then the subsequent check is also an indicator of your incorrect work. therefore it stands separately and in the future it will be easier to remove it.

but if you do not agree with my reasoning, I will remove it

Reply:
It should probably not be removed. Does one of your test cases cover this? This would help me to better understand what is going on.

Reply:
it's a good idea.
I added a simple example of `functionWork1_`, which the developer knows for sure that his return number is greater than 10. This does not give him a guarantee that the code will work correctly, but it improves it.

Reply:
Thanks.